### DISCUSSION TITLE
Fail2ban misses bots to ban

### DISCUSSION THREAD

**Discussion by ghost** [fail2ban/fail2ban#3647] [Q&A] [answered]
Hello. I have nginx-botsearch and nginx-bad-request filters enabled to get rid of bots on my website. But I still see a bot scanning for phpMyAdmin (I don't have that, my site is static), and nginx-botsearch filter does nothing. 
Log entry: `120.76.245.24 - - [09/Dec/2023:20:46:12 +0000] "GET http://172.188.76.121:80/php-myadmin/scripts/setup.php HTTP/1.0" 404 153 "-" "-"`
My jail.local:
```
[sshd]
bantime = 86400
action = %(action_)s
         %(action_abuseipdb)s[abuseipdb_apikey="my-key", abuseipdb_category="18,22"]
[nginx-botsearch]
enabled = true
maxretry = 3
findtime = 60
bantime = 86400
action = %(action_)s
         %(action_abuseipdb)s[abuseipdb_apikey="my-key", abuseipdb_category="19"]
[nginx-bad-request]
enabled = true
maxretry = 1
findtime = 60
bantime = 86400
action = %(action_)s
         %(action_abuseipdb)s[abuseipdb_apikey="my-key", abuseipdb_category="19"]
```
Can someone assist me in fixing this? Thank you.

**Comment by sebres** [comment, marked as answer]
Filter `nginx-botsearch` searching for the bots (so it'd match only certain agents in the message).
Filter `nginx-bad-request` searching for bad requests only (so bots scanning for ports and services, e. g. trying something other than HTTP proto or some injection, so caused response code `400` on nginx side).

If you want to ban 404 (and other failed) requests too, you can rewrite failregex of `nginx-bad-request` like here or create new jail with that:
```ini
failregex = ^<ADDR> - \S+ \[\] "[^"]*" 40\d
```
(just ensure your web-site doesn't generate 404 too, in order to avoid false positives, and/or probably increase `maxretry` additionally).
Although I don't understand why the message of logged request contains full URL, so with `http://.../` inside, whereas nginx basically logging only relative URI as I know.

Also take a look at https://github.com/fail2ban/fail2ban/issues/3518#issuecomment-1571761488 for similar filter (where you can define your own sub RE, e. g. define `_req_uri = php-myadmin|admin|other-uri-to-block` to capture certain URIs only). 
However in your case (because of strange full URL inside) it'd be rather something like this:
```ini
[Definition]
_req_method = [A-Z]{3,}
# _req_code = 4(?:0[034]|44)
_req_code = (?!401)[345]0\d
# _req_uri = \S+
_req_uri = php-myadmin|admin|other-uri-to-block
failregex = ^<HOST> -[^"]*\"%(_req_method)s\s+(?:[a-zA-Z]{3,10}://[^/]+)?/(?:%(_req_uri)s)\b\S* HTTP/[^"]+" (?:%(_req_code)s)\s
```
(where `(?:[a-zA-Z]{3,10}://[^/]+)?` would match `http://...`).

Also note similar discussion in #2583 about possible issues with that approach.
